### ISSUE TITLE
improve QGIS to handle different layer definition sections 

### ISSUE THREAD

**Issue by haubourg** [qwat/QWAT#209]
Currently all layer properties sections are mixed into one style, with many styles possible for a layer. 

This is prone to useability caveats, like switching styles and not having the same form or layer actions, 
and makes it harder to maintain, if one need to have the same forms but allow differents style switches. 
Proposal would be to have one default section for forms, actions etc.. and independant style (symbology + labels+ display field) switching. 
Being able to store several actions - forms should be preserved but triggered with explicity buttons. 

https://issues.qgis.org/issues/15733

**Comment by haubourg** [comment]
Discussed in PSC 2017 11 23:

- QLR is a whole layer definition extract with all "styles" 
- QML exports only the currently displayed "style" which is composed of labels + forms + styles etc..

What is expected is:

- add subentries in the "style" to save only forms / labels / styles / metadata etc..
- display the style title outside of the style widget, so that users know permanently what style is being used. 
- store a name in exported QML and give feeback when loading a new style (a qgs messagebar for instance) + rules for ensuring unique style naming

**Comment by haubourg** [comment]
This feature is meant to adress easier local customization of projects and easier resync with master project

**Comment by m-kuhn** [comment]
If you are ok with this proposal, let me know and I'll open a PR upstream

**Comment by haubourg** [comment]
@m-kuhn we first need an estimate for that work. Could you do that?

**Comment by m-kuhn** [comment]
2 hours?

**Comment by haubourg** [comment]
with doc, tests and bugfix included? Better count one day  ?

**Comment by haubourg** [comment]
If you agree, GO!

**Comment by m-kuhn** [comment]
Sounds good to be on the safe side.

I'll check if tests make sense.

**Comment by m-kuhn** [comment]
Sorry @haubourg the proposal here was only for the screenshot above, not for the full control over what to save. I have mixed this up with #238 .

The full proposal will require 2 - 3 days, sorry for not realizing that these are two different tickets earlier.

**Comment by haubourg** [comment]
Oh, OK, 2 hours sounded fast yes !

**Comment by m-kuhn** [comment]
https://github.com/qgis/QGIS/pull/5900

**Comment by haubourg** [comment]
@ponceta @dsavary
 Given this is a not something users will discuss, but usefull to GIS administrators, what is the priority of this task for 2018? My opinion is that it is very much needed and as a very high mutualization value for the whole QGIS community. I would then plead to launch that as soon as 3.0 is out.  
Opinions?

**Comment by dsavary** [comment]
@haubourg
It is a priority need for us . (Gis admin)
To plan as soon as possible ( 2018 )

**Comment by m-kuhn** [comment]
Highly appreciated also from a qgis community perspective.

**Comment by haubourg** [comment]
Agreed with the priority here. @ponceta do you give us a GO or do you need QWAT's team approval?

**Comment by dsavary** [comment]
Is this issue also to fix the bug with QLR export (Layer Definition File)  in QGIS 2.18  ?
The settings of the form are lost (only for the current style) 
I create a new issue for this ? 

Tested in QGIS 3.0 : the export QLR seems to be working well.

Bug demo gif : 
![export_qlr_lost_form](https://user-images.githubusercontent.com/6229296/38621826-398abeb0-3d91-11e8-94c0-cd34a5b8e20f.gif)

**Comment by haubourg** [comment]
I just launched a QEP for advanced proposal of export / import of QML parts here https://github.com/qgis/QGIS-Enhancement-Proposals/issues/125

Any comments welcome

**Comment by haubourg** [comment]
@ponceta @dsavary @kandre are you happy with the current state of the QEP?

**Comment by dsavary** [comment]
@haubourg very good job. 
I don't understand all the details about the discussion with Mathias Kuhn...
Perhaps, it may be necessary to specify, if the export of sections (symbole, form, ...) is done by default on the current style or the user can choose (before) in the menu one of the styles that he wants to export.
